features/285-design-webhooks/design.md

```diff
@@ -1,129 +1,324 @@
+The events would be delivered to web-hook handlers as an HTTP POST request to
+the handler endpoint.
 
-## Request Payload
+The handler endpoint can be specified as full URL or an absolute path. Absolute
+paths will be resolved to full URL using an inferred URL scheme and authority:
+- from HTTP request URL if the event is generated from it; otherwise,
+- from the tenant configuration.
 
-Each event type has a payload which carries relevant information, and each payload data is defined by each gear.
+Each event type can have multiple handlers; the order of deliveries is
+unspecified for AFTER events, and same order as in `skygear.yaml` for BEFORE
+events.
 
-All event payloads follows following payload format:
+BEFORE events will always be delivered before AFTER events. BEFORE events will
+be delivered in well-defined order during a request. AFTER events will be
+delivered in unspecified order.
 
-```json=
+Web-hook handler should be idempotent, since the same event may be delivered
+multiple times due to retries.
+
+Web-hook handler must return a status code within the 2XX range. Responding with
+status code outside the range, including 3XX & 5XX, would be considered a
+failed delivery.
+
+### BEFORE Events
+
+BEFORE events would be delivered to web-hook handlers synchronously, right
+before committing the result of the operation into database.
+
+Web-hook handler should respond with a JSON-formatted body to indicated whether
+the operation should be failed, for example:
+```json
+// Allowed
+{
+    "is_allowed": true
+}
+
+// Disallowed
 {
-    "event": "event_name",
+    "is_allowed": false,
+    "reason": "the metadata does not match the required format.",
     "data": {
-        "key_name_1": "event_data_key_value",
-        "key_name_2": "event_data_key_value",
-        ...,
-    },
-    "context": {
-        "user": <current user object>,
-        "req": {
-            "path": "request_path",
-            "body": <original_request_body>,
-            "id": "request_id"
+        "email": "invalid email format"
+    }
+}
+```
+If the operation is disallowed, a non-empty reason must be provided. Optional
+additional information can be included. If any of the delievery responses
+disallowed the operation, the operation is considered as failed, with the
+disallowing reasons and additional information as part of the error. For example:
+```json
+{
+    "error": {
+        "name": "WebHookError",
+        "code": 10000,
+        "message": "Operation is disallowed by web-hook",
+        "info": {
+            "errors": [
+                {
+                    "reason": "the metadata does not match the required format.",
+                    "data": {
+                        "email": "invalid email format"
+                    }
+                }
+            ]
         }
     }
 }
 ```
 
-Note that, `context` is above format is still under discussion.
+The time spent in a BEFORE event delivery must not exceed 5 seconds, otherwise
```

Comment:
Move the sentence `This timeout can be configured in app config` to the end of this paragraph. Rewrite it to `Both timeout are configurable in app config`.